elastic/kibana
# [Security Solution][CTI] Rule Preview backend update (introduces /preview endpoint)

## Summary

This is the backend portion of the Rule Preview implementation upgrade for all rules. Users can now experience the preview feature with results that are closely resembling the alerts that they would see if they were to execute the rules in conjunction with the alerting framework. 

The PR consists of the following changes:
- the creation of a /preview endpoint (where a POST request with rule params and an "invocation count" triggers the rule executors to find hits, converts them into "signals", and writes them to the preview index with a unique preview id, in a space aware fashion
- the refactor of ruleSO, where savedObjectsClient is no longer used to fetch a ruleSO, and instead a completeRule object is built out of SanitizedConfig, ruleParams, and an alertId (in the case of preview, a previewId)
- the creation of a /preview/index endpoint (where a POST request creates the `.siem-preview-signals` index if it doesn't exist.
- the refactor of driftTolerance where some redundant code has been removed.
- the addition of a preview script where developers can test the preview endpoint easily.

## Testing 

- delete your preview index (likely `.siem-preview-signals-default-000001`) if you have it
- visit the create rule page
- confirm that you now have a preview index (`.siem-preview-signals-default-000001`) with `.siem-signals` mappings
- visit the create rule page again and open Chrome Developer tools Network tab
- fill in all of the rule details until right before clicking the "Create & activate rule" button
- clear the Network tab and click the button
- right click on the first request (to `rules`) and copy it as cURL
- open a new terminal window and paste
- edit the request url to be `http://localhost:5601/api/detection_engine/rules/preview` by adding `/preview` at the end of the url.
- edit the payload of the request (at the very end, likely next to `"enabled":true,"throttle":"no_actions"`) by adding `"invocationCount":100` (or any number of times you would like to invoke the rule)
- confirm that the response on the terminal returns a unique preview id, and errors / warnings if there are any
- view the contents of the `.siem-preview-signals` index to confirm that the expected preview-items appear as expected.

you can also use the rule preview script at `x-pack/plugins/security_solution/server/lib/detection_engine/scripts/post_rule_preview.sh`, which would require you to create `custom-events` and `custom-threats` indices beforehand with data matching on `file.hash.md5` / `threat.indicator.file.hash.md5`.  Note: I am getting an unexpected error around `./check_env_variables.sh` not being present even though I can see it in the correct location, so I assume that there is a problem with my setup and not the script, I would be happy to know if you can run the script without any problems.

### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/master/RISK_MATRIX.mdx) |


### For maintainers
